url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/util/Graph.java

```diff
@@ -0,0 +1,261 @@
+    }
+
+    public Set getOutbounds(Object node) {
+        return (Set) outbounds.get(node);
+    }
+
+    public void clearOutbounds(Object node) {
+        Set s = (Set) outbounds.get(node);
+        if (s != null) {
+            Iterator i = s.iterator();
+            while (i.hasNext())
+                remove((Edge) i.next());
+        }
+    }
+
+    public void clearInbounds(Object node) {
+        Set s = (Set) inbounds.get(node);
+        if (s != null) {
+            Iterator i = s.iterator();
+            while (i.hasNext())
+                remove((Edge) i.next());
+        }
+    }
+
+    public void remove(Object node) {
+        nodes.remove(node);
+        clearInbounds(node);
+        clearOutbounds(node);
+    }
+
+    public void add(Edge e) throws GraphCycleException {
+        checkCycle(e);
+        nodes.add(e.source);
+        nodes.add(e.target);
+        edges.add(e);
+        Set in = (Set) inbounds.get(e.target);
+        if (in == null)
+            inbounds.put(e.target, in = new HashSet());
+        in.add(e);
+        Set out = (Set) outbounds.get(e.source);
+        if (out == null)
+            outbounds.put(e.source, out = new HashSet());
+        out.add(e);
+    }
+
+    public void checkCycle(Edge e) throws GraphCycleException {
+        HashSet visited = new HashSet();
+        visited.add(e.source);
+        checkCycle(e, visited);
+    }
+
+    private void checkCycle(Edge e, HashSet visited) throws GraphCycleException {
+        if (visited.contains(e.target)) {
+            throw new GraphCycleException("Circular reference found: " +
+                    e.source + " - " + e.target);
+        }
+        visited.add(e.target);
+        Set out = (Set) outbounds.get(e.target);
+        if (out != null) {
+            Iterator i = out.iterator();
+            while (i.hasNext()) {
+                checkCycle((Edge) i.next(), visited);
+            }
+        }
+    }
+
+    public void remove(Edge e) {
+        edges.remove(e);
+        Set in = (Set) inbounds.get(e.target);
+        if (in != null)
+            in.remove(e);
+        Set out = (Set) outbounds.get(e.source);
+        if (out != null)
+            out.remove(e);
+    }
+
+    public void sort() {
+        ordered = new ArrayList();
+        traversed = new HashSet();
+        Iterator i = nodes.iterator();
+        Set remains = new HashSet(nodes);
+
+        // First traverse nodes without inbounds
+        while (i.hasNext()) {
+            Object o = i.next();
+            Set in = (Set) inbounds.get(o);
+            if (in == null || in.isEmpty()) {
+                traverse(o);
+                remains.remove(o);
+            }
+        }
+
+        // Now traverse the rest
+        i = remains.iterator();
+        while (i.hasNext()) {
+            Object o = i.next();
+            if (!traversed.contains(o)) {
+                traverse(o);
+            }
+        }
+    }
+
+    private void traverse(Object node) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 150-186 into a separate method.